test/distributed/_composable/test_composability/test_3d_composability.py

```diff
@@ -0,0 +1,565 @@
+        if cpu_offload:
+            fsdp_config["offload_policy"] = CPUOffloadPolicy()
+
+        for layer_id in range(len(model)):
+            transformer_block = model[layer_id]
+            if pp_enabled:
+                # For PP, do not reshard after forward to avoid per-microbatch
+                # all-gathers, which can be expensive and non-overlapped
+                reshard_after_forward = False
+            else:
+                # As an optimization, do not reshard after forward for the last
+                # transformer block since FSDP would prefetch it immediately
+                reshard_after_forward = int(layer_id) < 7
+            fully_shard(
+                transformer_block,
+                **fsdp_config,
+                reshard_after_forward=reshard_after_forward,
+            )
+        fully_shard(model, **fsdp_config, reshard_after_forward=not pp_enabled)
+
+    def clip_grad_norm_(
+        self,
+        parameters: Union[torch.Tensor, Iterable[torch.Tensor]],
+        max_norm: float,
+        norm_type: float = 2.0,
+        error_if_nonfinite: bool = False,
+        foreach: Optional[bool] = None,
+        pp_mesh: Optional[DeviceMesh] = None,
+    ) -> torch.Tensor:
+        """
+        Clip the gradient norm of an iterable of parameters.
+
+        Gradient norm clipping requires computing the gradient norm over the entire model.
+        `torch.nn.utils.clip_grad_norm_` only computes gradient norm along DP/FSDP/TP dimensions.
+        We need to manually reduce the gradient norm across PP stages.
+        See https://github.com/pytorch/torchtitan/issues/596 for details.
+        """
+        grads = [p.grad for p in parameters if p.grad is not None]
+        total_norm = torch.nn.utils.get_total_norm(
+            grads, norm_type, error_if_nonfinite, foreach
+        )
+
+        if pp_mesh is not None:
+            if isinstance(total_norm, DTensor):
+                # will reach here if PP + other parallelism is used. If only using PP, total_norm will be a local tensor
+
+                # if total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`
+                # we can simply reduce the DTensor to get the total norm in this tensor's process group
+                # and then convert it to a local tensor
+                total_norm = total_norm.full_tensor()
+
+            if math.isinf(norm_type):
+                dist.all_reduce(
+                    total_norm, op=dist.ReduceOp.MAX, group=pp_mesh.get_group()
+                )
+            else:
+                total_norm **= norm_type
+                dist.all_reduce(
+                    total_norm, op=dist.ReduceOp.SUM, group=pp_mesh.get_group()
+                )
+                total_norm **= 1.0 / norm_type
+
+        torch.nn.utils.clip_grads_with_norm_(parameters, max_norm, total_norm, foreach)
+        return total_norm
+
+
+@dataclass
+class ParallelDims:
```

Comment:
I suspect we do not need to copy this class from TorchTitan.  It is a nice helper but in the unit test case you can probably just construct your mesh directly.